pkg/datasource/data.go

```diff
@@ -0,0 +1,420 @@
+	return ds.name
+}
+
+func (ds *dataSource) Type() Type {
+	return ds.dType
+}
+
+func (ds *dataSource) NewData() Data {
+	return ds.dPool.Get().(Data)
+}
+
+func (ds *dataSource) ByteOrder() binary.ByteOrder {
+	return ds.byteOrder
+}
+
+func resolveNames(id uint32, fields []*field, parentOffset uint32) (string, error) {
+	if id >= uint32(len(fields)) {
+		return "", errors.New("invalid parent id")
+	}
+	out := ""
+	if FieldFlagHasParent.In(fields[id].Flags) {
+		p, err := resolveNames(fields[id].Parent-parentOffset, fields, parentOffset)
+		if err != nil {
+			return "", errors.New("parent not found")
+		}
+		out = p + "."
+	}
+	out += fields[id].Name
+	return out, nil
+}
+
+// AddStaticFields adds a statically sized container for fields to the payload and returns an accessor for the
+// container; if you want to access individual fields, get them from the DataSource directly
+func (ds *dataSource) AddStaticFields(size uint32, fields []Field) (FieldAccessor, error) {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	idx := ds.payloadCount
+
+	// temporary write to newFields to not write to ds.fields in case of errors
+	var newFields []*field
+
+	parentOffset := len(ds.fields)
+
+	for _, f := range fields {
+		if _, ok := ds.fieldMap[f.FieldName()]; ok {
+			return nil, fmt.Errorf("field %q already exists", f.FieldName())
+		}
+		staticField, ok := f.(StaticField)
+		fieldName := f.FieldName()
+		if !ok {
+			return nil, fmt.Errorf("field %q is not statically sized or does not implement StaticField", fieldName)
+		}
+		nf := &field{
+			Name:         fieldName,
+			Index:        uint32(len(ds.fields) + len(newFields)),
+			PayloadIndex: idx,
+			Flags:        FieldFlagStaticMember.Uint32(),
+		}
+		nf.Size = staticField.FieldSize()
+		nf.Offs = staticField.FieldOffset()
+		if nf.Offs+nf.Size > size {
+			return nil, fmt.Errorf("field %q exceeds size of container (offs %d, size %d, container size %d)", nf.Name, nf.Offs, nf.Size, size)
+		}
+		if s, ok := f.(TypedField); ok {
+			nf.Kind = s.FieldType()
+		}
+		if tagger, ok := f.(TaggedField); ok {
+			nf.Tags = tagger.FieldTags()
+		}
+		if s, ok := f.(FlaggedField); ok {
+			nf.Flags |= uint32(s.FieldFlags())
+		}
+		if s, ok := f.(ParentedField); ok {
+			parent := s.FieldParent()
+			if parent >= 0 {
+				nf.Parent = uint32(parent + parentOffset)                          // TODO: validate?
+				nf.Flags |= FieldFlagHasParent.Uint32() | FieldFlagHidden.Uint32() // default to hide subfields
+			}
+		}
+		newFields = append(newFields, nf)
+	}
+
+	var err error
+	for i, f := range newFields {
+		f.FullName, err = resolveNames(uint32(i), newFields, uint32(parentOffset))
+		if err != nil {
+			return nil, fmt.Errorf("resolving full fieldnames: %w", err)
+		}
+	}
+
+	ds.fields = append(ds.fields, newFields...)
+
+	for _, f := range newFields {
+		ds.fieldMap[f.Name] = f
+	}
+
+	ds.payloadCount++
+
+	return &fieldAccessor{ds: ds, f: &field{
+		PayloadIndex: idx,
+		Size:         size,
```

Comment:
Shouldn't it set the static flag here? Otherwise the length checks won't work.

Reply:
No, it should get the Container flag (since it's the container for all static members) and have a separate bounds check for that flag in the `Set()` method; forgot that when changing from the `.Size() > 0` to the more explicit `StaticMember` flag. Added it (which also made your tests pass.)